src/protocol/boolean/bit_decomposition.rs

```diff
@@ -0,0 +1,210 @@
+use super::bitwise_lt::BitwiseLessThan;
+use super::bitwise_sum::BitwiseSum;
+use crate::error::Error;
+use crate::ff::{Field, Int};
+use crate::protocol::boolean::local_secret_shared_bits;
+use crate::protocol::boolean::solved_bits::SolvedBits;
+use crate::protocol::context::{Context, SemiHonestContext};
+use crate::protocol::reveal::Reveal;
+use crate::protocol::RecordId;
+use crate::secret_sharing::Replicated;
+
+/// This is an implementation of "3. Bit-Decomposition" from I. Damgård et al..
+///
+/// It takes an input `[a] ∈ F_p` and outputs its bitwise additive share
+/// `[a]_B = ([a]_0,...,[a]_l-1)` where `[a]_i ∈ F_p`.
+///
+/// 3. Bit-Decomposition
+/// "Unconditionally Secure Constant-Rounds Multi-party Computation for Equality, Comparison, Bits, and Exponentiation"
+/// I. Damgård et al.
+pub struct BitDecomposition {}
+
+impl BitDecomposition {
+    #[allow(dead_code)]
+    pub async fn execute<F: Field>(
+        ctx: SemiHonestContext<'_, F>,
+        record_id: RecordId,
+        a_p: &Replicated<F>,
+    ) -> Result<Vec<Replicated<F>>, Error> {
+        // step 1 in the paper is just describing the input, `[a]_p` where `a ∈ F_p`
+
+        // Step 2. Generate random bitwise shares
+        let r = SolvedBits::execute(ctx.narrow(&Step::GenerateRandomBits), record_id).await?;
+        // Would like to do something like below. Where should I add the `RandomBitsGenerator`?
+        // let r = ctx.random_bits_generator.take_one().await?;
+        let r = r.expect("SolvedBits aborted");
+
+        // Step 3, 4. Reveal c = [a - b]_p
+        let c = ctx
+            .narrow(&Step::RevealAMinusB)
+            .reveal(record_id, &(a_p - &r.b_p))
+            .await?;
+        let c_b = local_secret_shared_bits(c.as_u128(), ctx.role());
+
+        // Step 5. Add back [b] bitwise. [d]_B = BitwiseSum(c, [b]_B) where d ∈ Z
+        //
+        // `BitwiseSum` outputs `l + 1` bits, so [d]_B is (l + 1)-bit long.
+        let d_b = BitwiseSum::execute(ctx.narrow(&Step::AddBtoC), record_id, &c_b, &r.b_b).await?;
+
+        // Step 6. p <? d
+        let p_b = local_secret_shared_bits(F::PRIME.into(), ctx.role());
+        let q_p = BitwiseLessThan::execute(ctx.narrow(&Step::IsPLessThanD), record_id, &p_b, &d_b)
+            .await?;
+
+        // Step 7. a bitwise scalar value `f_B = bits(2^l - p)`
+        let l = F::Integer::BITS;
+        let x = 2_u128.pow(l) - F::PRIME.into();
```

Comment:
I'm trying to figure out if this is another time to use just the number of bits necessary to represent the field values (i.e. 5 bits for Fp31, not 8)